src/Integration/Telemetry/TelemetryManager.cs

```diff
@@ -163,36 +152,66 @@ private void DisableAllEvents()
-            KnownUIContexts.SolutionBuildingContext.UIContextChanged -= this.OnAnalysisRun;
-            KnownUIContexts.SolutionExistsAndFullyLoadedContext.UIContextChanged -= this.OnAnalysisRun;
+            this.knownUIContexts.SolutionBuildingContextChanged -= this.OnAnalysisRun;
+            this.knownUIContexts.SolutionExistsAndFullyLoadedContextChanged -= this.OnAnalysisRun;
         }
 
-        public void OptIn()
+        private void EnableAllEvents()
         {
-            this.telemetryRepository.Data.IsAnonymousDataShared = true;
-            this.telemetryRepository.Save();
+            this.tryUploadDataTimer.Elapsed += OnTryUploadDataTimerElapsed;
+            this.tryUploadDataTimer.Start();
 
-            EnableAllEvents();
+            this.firstCallDelayer.Elapsed += OnTryUploadDataTimerElapsed;
+            this.firstCallDelayer.Start();
+
+            this.knownUIContexts.SolutionBuildingContextChanged += this.OnAnalysisRun;
+            this.knownUIContexts.SolutionExistsAndFullyLoadedContextChanged += this.OnAnalysisRun;
         }
 
-        public async void OptOut()
+        private TelemetryPayload GetPayload()
         {
-            this.telemetryRepository.Data.IsAnonymousDataShared = false;
-            this.telemetryRepository.Save();
+            var numberOfDaysSinceInstallation = (long)(DateTime.Now - this.telemetryRepository.Data.InstallationDate).TotalDays;
+            return new TelemetryPayload
+            {
+                SonarLintProduct = "SonarLint Visual Studio",
+                SonarLintVersion = SonarLintVersion,
+                VisualStudioVersion = VisualStudioVersion,
+                NumberOfDaysSinceInstallation = numberOfDaysSinceInstallation,
+                NumberOfDaysOfUse = this.telemetryRepository.Data.NumberOfDaysOfUse,
+                IsUsingConnectedMode = this.solutionBindingTracker.IsActiveSolutionBound
+            };
+        }
 
-            DisableAllEvents();
+        private void OnAnalysisRun(object sender, UIContextChangedEventArgs e)
+        {
+            if (!e.Activated)
+            {
+                return;
+            }
 
-            await this.telemetryClient.OptOut(GetPayload());
+            var lastAnalysisDate = this.telemetryRepository.Data.LastSavedAnalysisDate;
+            if (lastAnalysisDate != DateTime.MinValue && (DateTime.Now - lastAnalysisDate).TotalDays < 1)
+            {
+                return;
+            }
+
+            this.telemetryRepository.Data.LastSavedAnalysisDate = DateTime.Now;
+            this.telemetryRepository.Data.NumberOfDaysOfUse++;
+            this.telemetryRepository.Save();
         }
 
-        public void Dispose()
+        private async void OnTryUploadDataTimerElapsed(object sender, ElapsedEventArgs e)
         {
-            DisableAllEvents();
+            var lastUploadDate = this.telemetryRepository.Data.LastUploadDate;
+            if (lastUploadDate != DateTime.MinValue && !IsNextDay(lastUploadDate))
```

Comment:
Move the first part of the check inside IsNextDay and rename it to ShouldUploadPayload or something like that. Even now it does more than just check if it is the next day.